src/devices/philips.c

```diff
@@ -0,0 +1,231 @@
+ * s - CRC: non-standard CRC-4, reverse-engineered using RevEng (4 bits)
+ *     http://reveng.sourceforge.net
+ *     width=4  poly=0x9  init=0x1  refin=false  refout=false  xorout=0x0  check=0x5  residue=0x0  name=(none)
+ *  
+ * Pulse width:
+ *      Short: 2000 us = 0
+ *      Long: 6000 us = 1
+ * Gap width:
+ *      Short: 6000 us
+ *      Long: 2000 us
+ * Gap width between packets: 29000 us
+ * 
+ * Presumably the 4-bit preamble is meant to be a sync of some sort, 
+ * but it has the exact same pulse/gap width as a short pulse, and
+ * gets processed as data.
+ *
+ * Copyright (C) 2017 Chris Coffey (kpuc@sdf.org)
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ */
+
+#include "rtl_433.h"
+#include "util.h"
+
+#define PHILIPS_BITLEN       112
+#define PHILIPS_PACKETLEN    4
+#define PHILIPS_STARTNIBBLE  0x0
+
+/* Map channel values to their real-world counterparts */
+static const uint8_t channel_map[] = { 2, 0, 1, 0, 3 };
+
+
+/* philips_crc4_bit():
+ * Compute CRC-4 for a byte sequence, one bit at a time.
+ *
+ * Note that Philips uses a custom CRC implementation for this device; see above
+ * for details.
+ * 
+ * This function was generated using crcany (https://github.com/madler/crcany)
+ * License information: "This code is under the zlib license, permitting free
+ * commercial use."
+ */
+static unsigned philips_crc4_bit(unsigned crc, void const *mem, size_t len) 
+{
+    unsigned char const *data = mem;
+    if (data == NULL)
+        return 0x1;
+    crc <<= 4;
+    while (len--) {
+        crc ^= *data++;
+        for (unsigned k = 0; k < 8; k++)
+            crc = crc & 0x80 ? (crc << 1) ^ 0x90 : crc << 1;
+    }
+    crc >>= 4;
+    crc &= 0xf;
+    return crc;
+}
+
+
+/* philips_crc4_rem():
+ * Compute CRC-4 of the remaining high bits in the low byte of val.
+ *
+ * Note that Philips uses a custom CRC implementation for this device; see above
+ * for details.
+ * 
+ * This function was generated using crcany (https://github.com/madler/crcany)
+ * License information: "This code is under the zlib license, permitting free
+ * commercial use."
+ */
+static unsigned philips_crc4_rem(unsigned crc, unsigned val, unsigned bits) 
+{
+    crc <<= 4;
+    val &= ((1U << bits) - 1) << (8 - bits);
+    crc ^= val;
+    while (bits--)
+        crc = crc & 0x80 ? (crc << 1) ^ 0x90 : crc << 1;
+    crc >>= 4;
+    crc &= 0xf;
+    return crc;
+}
+
+
+static int philips_callback(bitbuffer_t *bitbuffer) 
+{
+    char time_str[LOCAL_TIME_BUFLEN];
+    uint8_t *bb;
+    unsigned int i;
+    uint8_t a, b, c;
+    uint8_t packet[PHILIPS_PACKETLEN];
+    uint8_t r_crc, c_crc;
+    uint8_t channel, battery_status;
+    int tmp;
+    float temperature;
+    data_t *data;
+
+    /* Get the time */
+    local_time_str(0, time_str);
```

Comment:
only call local_time_str after the last early return, preferably just before data_make

Reply:
The debug statements in the validation checks use the time from local_time_str() as well -- do you not want the time in those? Or, I could call local_time_str() twice, once before the early returns, and again just before data_make()

Reply:
Since it's merged already consider it for future work where applicable. The reasoning is that we generally don't want all the decoders to do expensive work where it's likely not a valid packet.

Reply:
Gotcha, thanks for the review.